### PR TITLE
Added KeyName option to Signature

### PR DESCRIPTION
**Pull request by hk0**
Sign method now accepts parameter: key_name = <string>, this will be
put in KeyInfo part of Signature

### CODE REVIEW

File: signxml/__init__.py [hunk 1]
```diff
@@ -315,6 +315,12 @@ def sign(self, method=methods.enveloped, algorithm="rsa-sha256", key=None, passp
         :param reference_uri:
             Custom reference URI to incorporate into the signature. Only used when ``method`` is set to ``detached``.
         :type reference_uri: string
+        :param key_name:
+            The KeyName to include in KeyInfo/KeyName signature. Selection to include in Keyinfo is as follows:
+            a) /KeyInfo/KeyName, when parameter ``key_name`` is supplied
+            b) /KeyInfo/X509Data/X509Certificate(s), when ``cert`` parameter is supplied
+            c) /KeyInfo/KeyValue/..., otherwise !
```
Comment (kislyuk) [inline signxml/__init__.py]: Please remove ` !` from the end of the line.

File: signxml/__init__.py [hunk 2]
```diff
@@ -389,9 +395,10 @@ def sign(self, method=methods.enveloped, algorithm="rsa-sha256", key=None, passp
                 signature = long_to_bytes(r).rjust(32, b"\0") + long_to_bytes(s).rjust(32, b"\0")
 
             signature_value.text = ensure_str(b64encode(signature))
-
             key_info = SubElement(self.sig_root, ds_tag("KeyInfo"))
-            if cert_chain is None:
+            if key_name is not None:
+                SubElement(key_info, ds_tag("KeyName")).text = key_name
+            elif cert_chain is None:
```
Comment (kislyuk) [inline signxml/__init__.py]: These conditions are not mutually exclusive. This has to be an `if`, not an `elif`, if I understand correctly.

**Comment by kislyuk** [comment]
Please remove the `test/.test.py.swp` file from the changeset.

File: signxml/__init__.py (hunk 2 above)
Comment (hk0) [inline signxml/__init__.py]: Hi Andrey, you are right, they are not mutually exclusive, according to
scheme this is definition:

 Schema Definition:

   <element name="KeyInfo" type="ds:KeyInfoType"/>
   <complexType name="KeyInfoType" mixed="true">
     <choice maxOccurs="unbounded">
       <element ref="ds:KeyName"/>
       <element ref="ds:KeyValue"/>
       <element ref="ds:RetrievalMethod"/>
       <element ref="ds:X509Data"/>
       <element ref="ds:PGPData"/>
       <element ref="ds:SPKIData"/>
       <element ref="ds:MgmtData"/>
       <any processContents="lax" namespace="##other"/>
       
     </choice>
     <attribute name="Id" type="ID" use="optional"/>
   </complexType>


But now the use-case is important: I do not want always to include
KeyInfo/KeyValue/X509Data
How to express for that ?

Would it be better to pass list of KeyInfo options to sign method ?

for example syntax:

...sign(... KeyInfo = { 'KeyName':<string>, 'KeyValue': <boolean>,
'X509Data': <boolean>} )

Sample call:
...sign(... KeyInfo = { 'KeyName':'MyName', 'KeyValue': False, 'X509Data':
False} )

What would be your preference ?

rgds Harry

2015-09-11 14:36 GMT+02:00 Andrey Kislyuk <notifications@github.com>:

**Comment by hk0** [comment]
Added: Mutual exists option to KeyInfo + KeyName + X509Data.

**Pull request closed without merging**